### ISSUE TITLE
:lexpr/:cexpr fail to populate the location/quickfix list in some cases

### ISSUE THREAD

**Issue by ghost** [vim/vim#569]
Currently, with Vim 7.4, patch 1141, using something like `:lexpr system("ls)` or `:cexpr system("grep -n xyz *)` fails if an errorformat is already defined for a filetype.

For example, if you are working with a C filetype, and have the following errorformat defined for use with `clang`:

``` vim
let &errorformat = '%E%f:%l:%c: fatal error: %m,' .
      \ '%E%f:%l:%c: error: %m,' .
      \ '%W%f:%l:%c: warning: %m,' .
      \ '%-G%\m%\%%(LLVM ERROR:%\|No compilation database found%\)%\@!%.%#,' .
      \ '%E%m'
```

running `:cexpr system("ls") | lopen` fails to populate the quickfix list. It will only work with something like `cexpr system("clang".expand("%"))` since the erroformat was defined for `clang`. However if Vim was invoked with no filetype argument (i.e. just `vim`), then using `:cexpr system("ls")` works as expected and creates a location list listing the directories in cwd. 

This seems like a regression to me. Currently, I cannot set a errorformat for a filetype and continue to use `:lexpr` and `:cexpr` to create location/quickfix lists for a different command, like I could before Vim 7.4 patch 1013.

This bug was introduced in Vim 7.4 patch 1013:

``` git
commit 9b05a0d0f94d8c4c1ddd51e7f31b73f7556bdbdc
Author: Bram Moolenaar <Bram@vim.org>
Date:   Thu Dec 31 21:19:49 2015 +0100

    patch 7.4.1013
    Problem:    The local value of 'errorformat' is not used for ":lexpr" and
                ":cexpr".
    Solution:   Use the local value if it exists. (Christian Brabandt) Adjust the
                help for this.
```

Reverting this change fixes the issue. 

Is there a way to prevent the use of the errorformat in this case? Or should this change be improved upon or reverted?

**Comment by brammool** [comment]
Akshay wrote:

> Currently, using something like `:lexpr system("ls)` or `:cexpr system("grep -n xyz *)` fails if an errorformat is already defined for a filetype
> 
> For example, if you are working with a C filetype, and have the following errorformat defined for use with `clang`:
> 
> ``` vim
> let &errorformat = '%E%f:%l:%c: fatal error: %m,' 
>       \ '%E%f:%l:%c: error: %m,' 
>       \ '%W%f:%l:%c: warning: %m,' 
>       \ '%-G%\m%\%%(LLVM ERROR:%\|No compilation database found%\)%\@!%%#,' 
>       \ '%E%m'
> ```
> 
> running `:cexpr system("ls") | lopen` fails to populate the quickfix list However if Vim was invoked with no filetype argument (ie just `vim`), then using `:cexpr system("ls")` works as expected and creates a location list listing the directories in cwd 
> 
> This bug was introduced in Vim 74 patch 1013:
> 
> ``` git
> commit 9b05a0d0f94d8c4c1ddd51e7f31b73f7556bdbdc
> Author: Bram Moolenaar <Bram@vimorg>
> Date:   Thu Dec 31 21:19:49 2015 +0100
> 
>     patch 741013
>     Problem:    The local value of 'errorformat' is not used for ":lexpr" and
>                 ":cexpr"
>     Solution:   Use the local value if it exists (Christian Brabandt) Adjust the
>                 help for this
> ```
> 
> Reverting this change fixes the issue 

So, is the problem caused by this patch bigger than what it solves?
I tend to think so.

Perhaps we can find another solution for what this patch was trying to
solve.

## 

Me?  A skeptic?  I trust you have proof.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by ghost** [comment]
I agree. 

I've run cases to find out what the patch was for, but I haven't found anything. 

Before patch 7.4.1013, I could have an error format defined for `python -m py_compile` and have a `lexpr` for flake8 (a python linter) like so: `:lexpr system("flake8 ".expand("%"))`. Even though flake8's error format seems to be different than Python's, it would still populate the location list correctly, and I could even jump through them with `:lnext` and `:lprev`. 

I could have both `:cexpr` for the above python command that would presumably use my errorformat setting designed for `python -m py_compile`, which would all work fine and highlight the quickfix list properly (i.e. I can navigate through the errors and no `||` appears indicating that quickfix list failed to parse errors) and `:lexpr` for `flake8` and it would still highlight the location list correctly (i.e. I can navigate through location list) without me having to set any other errorformat designed for `flake8`.

**Comment by chrisbra** [comment]
I think the patch started out from this discussion: https://groups.google.com/d/msg/vim_use/UAC54sqT1G4/cIxehPEnBAAJ
However already in that discussion @lcd047 mentioned that this would break current scripts such as syntastic. He then explained it again when the patch was released https://groups.google.com/d/msg/vim_dev/qFWC-08BsPs/Ga_kPL5yAgAJ

As I said earlier, I am fine with either behaviour. I just thought that not using the local errorformat value would be inconsistent and unexpected.

**Comment by ghost** [comment]
Hi Christian/Bram,

Personally, I think it is unexpected that `:lexpr` and `:cexpr` don't work as described in Vim's documentation (See `:help :cexpr`) when another command is run that the errorformat was not designed for. 

For example, the doc says I can use `:cexpr system("grep -n xyz *")`, however, I am not able to do this if a filetype plugin already has an errorformat.

That said, I always thought it would be better to have two errorformat settings. So for example, you can have an 'errorformat' setting and retain the current behavior of `:cexpr`, but then introduce a new 'lerrorformat' or similar (sort of how mapleader and localleader work).  Another option is to have `:cexpr` read the errorformat setting but `:lexpr` should not by default. Those are my two cents.

Thanks.

**Comment by ghost** [comment]
Also, if anyone wants, here is the patch to revert 1013:

``` patch

diff --git a/runtime/doc/quickfix.txt b/runtime/doc/quickfix.txt
index 674470e..9858d4e 100644
--- a/runtime/doc/quickfix.txt
+++ b/runtime/doc/quickfix.txt
@@ -1,4 +1,4 @@
-*quickfix.txt*  For Vim version 7.4.  Last change: 2015 Dec 31
+*quickfix.txt*  For Vim version 7.4.  Last change: 2016 Jan 19


          VIM REFERENCE MANUAL    by Bram Moolenaar
@@ -217,9 +217,9 @@ command with 'l'.
 :cex[pr][!] {expr} Create a quickfix list using the result of {expr} and
            jump to the first error.
            If {expr} is a String, then each new-line terminated
-           line in the String is processed using the value
-           of 'errorformat' (buffer-local value if it was set)
-           and the result is added to the quickfix list.
+           line in the String is processed using the global value
+           of 'errorformat' and the result is added to the
+           quickfix list.
            If {expr} is a List, then each String item in the list
            is processed and added to the quickfix list.  Non
            String items in the List are ignored.
diff --git a/src/quickfix.c b/src/quickfix.c
index 21ac1c0..014b93d 100644
--- a/src/quickfix.c
+++ b/src/quickfix.c
@@ -4230,10 +4230,7 @@ ex_cexpr(eap)
    if ((tv->v_type == VAR_STRING && tv->vval.v_string != NULL)
        || (tv->v_type == VAR_LIST && tv->vval.v_list != NULL))
    {
-       char_u *efm = *curwin->w_buffer->b_p_efm == NUL ? p_efm
-                         : curwin->w_buffer->b_p_efm;
-
-       if (qf_init_ext(qi, NULL, NULL, tv, efm,
+       if (qf_init_ext(qi, NULL, NULL, tv, p_efm,
                (eap->cmdidx != CMD_caddexpr
                 && eap->cmdidx != CMD_laddexpr),
                 (linenr_T)0, (linenr_T)0, *eap->cmdlinep) > 0
```

**Comment by ghost** [comment]
Closing this since it was resolved in commit d6357e8f93c5 (vim 7.4.1149)

Thanks Bram and Christian!
